Fixed Update Configuration to uncomment commented out Attribute when found

aroberson opened the pull request.

Commit:
Fixed Update Configuration to uncomment commented out Attribute when found

gavin-scott commented on src/main/java/com/dell/isg/smi/service/server/configuration/manager/ConfigurationManagerImpl.java:
```diff
@@ -307,51 +314,37 @@ private SHARE_TYPES getShareTypeEnum(int type) {
-
-				SystemConfiguration systemConfig = (SystemConfiguration) binder.unmarshal(document);
-
-				updatedServerComponents = updateXMLSystemConfiguration(systemConfig, request);
```
Looks like you can delete `updateXMLSystemConfiguration` now, not used anywhere

rahman-muhammad replied:
Yes, updateXMLSystemConfiguration and other filter function will go away

gavin-scott commented on src/main/java/com/dell/isg/smi/service/server/configuration/manager/ConfigurationManagerImpl.java:
```diff
@@ -307,51 +314,37 @@ private SHARE_TYPES getShareTypeEnum(int type) {
-				SystemConfiguration systemConfig = (SystemConfiguration) binder.unmarshal(document);
-
-				updatedServerComponents = updateXMLSystemConfiguration(systemConfig, request);
-
```
Also you can delete `ComponentPredicate.updateXMLSystemConfiguration` and `ComponentPredicate.filterRequestServerComponents`

gavin-scott commented on src/test/java/com/dell/isg/smi/service/server/configuration/manager/ConfigurationManagerImplTest.java:
```diff
@@ -0,0 +1,219 @@
+        ConfigurationManagerImpl impl = new ConfigurationManagerImpl();
+        // TODO: figure out how to get this auto-wired
+        impl.setComponentPredicate(new ComponentPredicate());
+        configManager = impl;
```
This isn't needed any more since `updateConfiguration` doesn't use `componentPredicate` anymore. So `setup()` and the private variable can be removed and just do `new ConfigurationManagerImpl()`

gavin-scott commented on src/test/java/com/dell/isg/smi/service/server/configuration/manager/ConfigurationManagerImplTest.java:
```diff
@@ -0,0 +1,219 @@
+        }
+    }
+
+    Node textToNode(DocumentBuilder documentBuilder, String text) throws IOException, SAXException {
```
Looks like everything here and below is just exploratory, `configManager` is not being called. Just delete since it's not actually testing code.

gavin-scott commented on src/test/resources/exported-scp-2.xml:
```diff
@@ -0,0 +1,17 @@
+<SystemConfiguration Model="PowerEdge C6320" ServiceTag="64BMW52" TimeStamp="Mon Aug 14 18:21:24 2017">
+
+ <Component FQDD="BIOS.Setup.1-1">
```
Remove this file since tests aren't using it.

gavin-scott commented on src/test/java/com/dell/isg/smi/service/server/configuration/manager/ConfigurationManagerImplTest.java:
```diff
@@ -0,0 +1,219 @@
+                            if (nested.getFQDD(). equals("Disk.Bay.0:Enclosure.Internal.0-0:RAID.Slot.1-1")) {
+                                for (Attribute attribute : nestedComponent.getAttributes()) {
+                                    if (attribute.getName().equals("RAIDHotSpareStatus")) {
+                                        assertTrue(attribute.getValue().equals("Yes"));
```
Is it worth adding a check that a commented-out subcomponent value can be updated?

aroberson replied:
There is not a commented out attribute at that level which is why i did not test it

gavin-scott replied:
RAID.Slot.1-1/Disk.Virtual.0:RAID.Slot.1-1 has e.g. T10PIStatus : https://github.com/aroberson/smi-service-dell-server-configuration-profile/blob/eee11adc42cd09974083a91d218a4babd6cb5229/src/test/resources/exported-scp-1.xml#L1715

gavin-scott reviewed (changes requested):
Request a few minor clean-ups. Otherwise looks great to me.

aroberson commented:
@gavin-scott Updated requested changes

gavin-scott commented:
👍

gavin-scott commented:
@muqeethkowkab / @michaelsteven would one of you be able to review this? @aroberson needed it for the workflow he is working on as well and was able to get the code working to set attributes that were commented out in the exported SCP XML file.

muqeethkowkab reviewed (changes requested):
Below are the changes required for this PR approval. 

1) Add a request paramater flag as we discuss prior to perform the force update (i.e. to uncomment the attributes and update). By default it should be the normal flow, only if flag is true then go with the uncomment logic.

2) Its is highly recommended to use PreviewImport prior to final import call in case of updating machine critical attributes. 

3) Also please update the Swagger and ReadMe Documents.

4) Please add the FIT test cases for the new scenarios. 

Otherwise, Uncommenting logic looks good to me.

muqeethkowkab commented:
Also there is merge conflict please resolve it.

rahman-muhammad commented:
@aroberson @gavin-scott , 
Current behavior for update components is not entertaining components which are marked as  "  <!--  comments  --> "  in XML configuration template and ignore to update target configuration XML file.
 This PR will improve the existing implementation and add all user provided components to the target configuration XML file to be applied on target server.  There is no harm in doing this way, it will take care of both commented and non-commented components,
I would highly recommend to add Boolean flag to the Rest API endpoint e.g. forceUpdate= true || false.   Consumer calling same API with forceUpdate=true, backend implementation should update all components irrespective of the commented/non-commented otherwise update only non-commented components, this will also ensure consumer is aware of the possibility that physical server may go into a bad state if things go wrong.  
This recommended change will mimic iDRAC behavior for similar cases e.g. implemented as 
Forced/Graceful shut down during firmware update, SCP configuration, OS deployment etc.

aroberson commented:
@rahmanmuhamad Conflicts have been resolved.  Added Query Param forceUpdate on updateComponents.  Also added unit test confirming uncomment only occurs on force Update true.

rahman-muhammad commented on src/main/java/com/dell/isg/smi/service/server/configuration/controller/ConfigurationController.java:
```diff
@@ -521,7 +520,9 @@
     @RequestMapping(value = "/updateComponents", method = RequestMethod.POST, headers = "Accept=application/json", consumes = "application/json", produces = "application/json")
     @ApiOperation(value = "Update Components", nickname = "updateComponents", notes = "This operation updates the server system configuration", response = ServiceResponse.class)
-    public ResponseEntity<ServiceResponse> updateComponents(@RequestBody @Valid ComponentList request, BindingResult bindingResult) throws Exception {
+    public ResponseEntity<ServiceResponse> updateComponents(@ApiParam(value = "ComponentList to Update in the Configuration", required = true) @RequestBody @Valid ComponentList request,
```
Can you add forceUpdate flag as a member variable to  ComponentList ?  doing  this way entire request will be passed in as body and  also simplify handling attribute handling at controller layer

muqeethkowkab reviewed (approved):
Changes Looks Good.
